Implement math-depth.

eerii opened the pull request:
#### 11f62061f97dc183ea06046b0b7e0d7065f80f6d
<pre>
Implement math-depth.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300967">https://bugs.webkit.org/show_bug.cgi?id=300967</a>

Reviewed by Antti Koivisto and Sam Weinig.

Parse and implement the `math-depth` property defined on MathML Core:
<a href="https://w3c.github.io/mathml-core/#propdef-math-depth.">https://w3c.github.io/mathml-core/#propdef-math-depth.</a>
This is gated behind the `CSSMathDepth` preference.
At the moment the property doesn&apos;t do anything, a followup patch for
`font-size: math` is required.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/parsing/math-depth-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/parsing/math-depth-valid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt: Removed.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/mathml.css:
(math):
(mfrac &gt; *):
(mroot &gt; :not(:first-child)):
(msub &gt; :not(:first-child),):
(mroot &gt; *:last-child): Deleted.
(msub &gt; * + *, msup &gt; * + *, msubsup &gt; * + *, mmultiscripts &gt; * + *, munder &gt; * + *, mover &gt; * + *, munderover &gt; * + *): Deleted.
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialMathDepth):
(WebCore::RenderStyle::mathDepth const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setMathDepth):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/values/math/StyleMathDepth.cpp: Added.
(WebCore::Style::CSSValueConversion&lt;MathDepth&gt;::operator):
* Source/WebCore/style/values/math/StyleMathDepth.h: Added.

Canonical link: <a href="https://commits.webkit.org/302116@main">https://commits.webkit.org/302116@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cf5016aba735c9e0dfc8b2e9f454572a85d84d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38788 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135335 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79491 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129834 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97420 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65341 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/73 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114632 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77987 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/83 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78643 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119993 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108435 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137818 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126423 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/99 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105949 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110980 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105685 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26957 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/81 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29545 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/52252 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/151 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159444 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/82 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39812 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/125 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/math/StyleMathDepth.cpp:
```diff
@@ -0,0 +1,53 @@
+
+#include "AnimationUtilities.h"
+#include "RenderStyleInlines.h"
+#include "StylePrimitiveNumericTypes+Blending.h"
```
This header include should not be needed.

weinig commented on Source/WebCore/style/values/math/StyleMathDepth.cpp:
```diff
@@ -0,0 +1,53 @@
+#include "config.h"
+#include "StyleMathDepth.h"
+
+#include "AnimationUtilities.h"
```
This header include should not be needed.

weinig commented on Source/WebCore/style/values/math/StyleMathDepth.cpp:
```diff
@@ -0,0 +1,53 @@
+    if (auto function = dynamicDowncast<CSSFunctionValue>(value); function && function->name() == CSSValueAdd) {
+        if (auto addFunction = requiredListDowncast<CSSFunctionValue, CSSPrimitiveValue>(state, value)) {
+            RefPtr first = addFunction->item(0);
+            auto value = first && first->isInteger() ? first->integerDeprecated() : 0;
```
Please don't introduce any new uses of the "Deprecated" functions. They have that suffix for a reason.

This should instead just use normal style resolution.

```
        if (auto addFunction = requiredListDowncast<CSSFunctionValue, CSSPrimitiveValue, 1>(state, value)) {
            Ref first = addFunction->item(0);
            auto value = toStyleFromCSSValue(state, first);
            return { state.parentStyle().mathDepth().value + value.value };
        }
```

eerii replied:
So sorry about that. It was a placeholder and I had a TODO to change it but I don't know where it went. Thank you for the suggestion.

weinig commented on Source/WebCore/style/values/math/StyleMathDepth.h:
```diff
@@ -0,0 +1,46 @@
+
+// <'math-depth'> = auto-add | add(<integer>) | <integer>
+// https://w3c.github.io/mathml-core/#propdef-math-depth
+struct MathDepth : Integer<> {
```
This could probably just be 

```
using MathDepth = Integer<>;
```

but this is fine as well.

eerii replied:
Huh, it seems that doing that there are crashes. Since we are overloading the template in `StyleMathDepth.cpp` I think having the same type may be causing some bad things. I reverted this to having a struct as before.

weinig commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -5845,6 +5845,26 @@
+                "animation-wrapper": "StyleTypeWrapper",
+                "settings-flag": "cssMathDepthEnabled",
+                "style-converter": "StyleType<MathDepth>",
+                "top-priority": true,
```
I'd like @anttijk to weigh in on what the costs of adding a new top-priority property might be.

eerii replied:
Great, thanks. For what is worth, Chromium does this too, since this property needs to resolve before the `font-size` one: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/css_properties.json5;l=1747?q=%22math-depth%22%20file:properties.json

fred-wang replied:
Yes, we have to do that, Stylo (used by Firefox/Servo) does that too: https://searchfox.org/firefox-main/rev/8e5d58cfed616cb90586c614e53d8ab1ffc8af27/servo/components/style/properties/data.py#55

In practice math-depth is only used for MathML subtrees, so that shouldn't impact the performance when MathML is not used.

anttijk replied:
I don't think there is really much cost.

weinig reviewed (approved):
Looks fine to me, but I like to hear @anttijk's thoughts on the top-priority part.

eerii committed:
Implement math-depth.
https://bugs.webkit.org/show_bug.cgi?id=300967

Reviewed by Antti Koivisto and Sam Weinig.

Parse and implement the `math-depth` property defined on MathML Core:
https://w3c.github.io/mathml-core/#propdef-math-depth.
This is gated behind the `CSSMathDepth` preference.
At the moment the property doesn't do anything, a followup patch for
`font-size: math` is required.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/parsing/math-depth-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/parsing/math-depth-valid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-fonts/math-script-level-and-math-style/math-script-level-001.tentative-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt: Removed.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/mathml.css:
(math):
(mfrac > *):
(mroot > :not(:first-child)):
(msub > :not(:first-child),):
(mroot > *:last-child): Deleted.
(msub > * + *, msup > * + *, msubsup > * + *, mmultiscripts > * + *, munder > * + *, mover > * + *, munderover > * + *): Deleted.
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialMathDepth):
(WebCore::RenderStyle::mathDepth const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setMathDepth):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/values/math/StyleMathDepth.cpp: Added.
(WebCore::Style::CSSValueConversion<MathDepth>::operator):
* Source/WebCore/style/values/math/StyleMathDepth.h: Added.

Canonical link: https://commits.webkit.org/302116@main
